Use different device tree with different media devices enabled

dsemenets opened the pull request.

dsemenets committed:
linux: Align field in domd device tree

Minor changes - align fields in bcm2712-raspberrypi5-domd.dts

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

lorc commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-usb.dts:
```diff
@@ -0,0 +1,262 @@
+	#address-cells = <0x2>;
+	#size-cells = <0x2>;
+
+	gic: gic {
```
You don't need this node. It is added automatically by xen tools/xenlib

GrygiriiS replied:
We need it to compile this dts - xen replaces it when loading

lorc replied:
Hmm, AFAIK, Xen provides `GUEST_PHANDLE_GIC`, which should be used like this:

https://github.com/xen-troops/meta-xt-prod-devel-rcar/blob/master/meta-xt-prod-devel-rcar-driver-domain/recipes-kernel/linux/files/r8a7795-salvator-x-4x2g-domd.dts#L15

https://github.com/xen-troops/meta-xt-prod-devel-rcar/blob/master/meta-xt-prod-devel-rcar-driver-domain/recipes-kernel/linux/files/r8a7795-salvator-x-4x2g-domd.dts#L20

I am surprised that it works without this

GrygiriiS replied:
So, As per my understanding this PHANDLE_GIC can be rolled back and `gic` node can be used for domD with dom0less boot. Right? @lorc @dsemenets @oleksiimoisieiev

lorc replied:
Yes. I am sorry for the fuss :(

 Didn't expected dom0less to work in **that** way

lorc commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-usb.dts:
```diff
@@ -0,0 +1,262 @@
+		mip0: msi-controller@1000130000 {
+			compatible = "brcm,bcm2712-mip-intc";
+			reg = <0x10 0x00130000 0x0 0xc0>;
+			interrupt-parent = <&gic>;
```
I am interested how it is working, given that xen provides special `phandle` for gic

lorc commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-usb.dts:
```diff
@@ -0,0 +1,262 @@
+				  0x43000000 0x4 0x0 0x1c 0x0 0x3 0x0>;
+			interrupt-parent = <&gic>;
+			/*
+			 * HACK: But the mip0: msi-controller@130000 doesn't have IRQs defined
```
I don't think we need this hack in the partial device tree

GrygiriiS replied:
no. I'd like it to stay here as there is no other place for it - and It clearly identifies WHY our DT is differ from original RPI5 DT.

lorc replied:
I can understand why do you need this in the Xen device tree. But why do you need this in DomD DTS?

GrygiriiS replied:
It's hack for DomD DT, as in dom0less mode it makes PCIe IRQ work properly.

lorc commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-usb.dts:
```diff
@@ -0,0 +1,262 @@
+			compatible = "brcm,brcmstb-reset";
+			#reset-cells = <0x1>;
+			reg = <0x10 0x1504318 0x0 0x30>;
+			extracted,path = "/axi/reset-controller@1504318";
```
What this property is used for?

GrygiriiS replied:
Yep this is leftover from imagebuilder.

dsemenets replied:
Should I remove this?

lorc replied:
Yes. It is not used by anyone

dsemenets replied:
fixed

lorc commented on recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:
```diff
@@ -11,14 +11,18 @@ TEMPLATE_FILE = "boot.cmd.xen.in"
 
+DOMD_MMC_DT_NAME = "${RPI_SOC_FAMILY}-${MACHINE}-mmc.dtb"
+DOMD_USB_DT_NAME = "${RPI_SOC_FAMILY}-${MACHINE}-usb.dtb"
+DOMD_BOOT_DEVICE = "${@bb.utils.contains('MACHINE_FEATURES', 'domdusb', '\\/dev\\/sda2', '\\/dev\\/mmcblk0p2', d)}"
```
I believe it is better to do via device tree overlays.

GrygiriiS replied:
Right. Possible Idea is to have mostly empty bcm2712-raspberrypi5-domd.dts (root, `gic` and `passthrough` with cells/ranges/compatible) +
```
 bcm2712-raspberrypi5-domd-mmc.dtso,
 bcm2712-raspberrypi5-domd-usb.dtso,
 bcm2712-raspberrypi5-domd-net.dtso
```
u-boot reads - bcm2712-raspberrypi5-domd.dts and applies -mmc|usb.dtso + -net.dtso depending on MACHINE_FEATURES.

There could be common parts of-course 

Super ideal result is to be able to do smth. like this:
`moulin X --ENABLE_DOMD_USB ----ENABLE_DOMD_NET`

dsemenets replied:
> Super ideal result is to be able to do smth. like this:

This already implemented. I have appropriate PR

dsemenets committed:
linux: Use device tree overlays for domd dt

Use device tree overlays for creating domd device tree.

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

dsemenets committed:
linux: Add dt overlays with mmc and usb device

Move mmc device from domd device tree to overlay. Add dt overlay with
usb device.

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

oleksiimoisieiev commented on recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in:
```diff
@@ -23,8 +23,5 @@ fdt set /chosen/domD/module@@@DOMD_IMG_ADDR@@ reg <0x0 @@DOMD_IMG_ADDR@@ 0x0 0x$
-setenv fdt_high 0xffffffffffffffff
-
-booti @@XEN_IMG_ADDR@@ - ${fdt_addr}
+fdt set /chosen/domD/module@@@DOMD_DTB_ADDR@@ reg <0x0 @@DOMD_DTB_ADDR@@ 0x0 0x6000 >
```
why did you replaced ${filesize} with 0x6000?

dsemenets replied:
Because we modify loaded DT and size if DT changed. There is no way to calculate accurate DT size

oleksiimoisieiev commented on recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.dtapply:
```diff
@@ -0,0 +1,3 @@
+fatload @@BOOT_MEDIA@@ 0 @@XEN_DTBO_ADDR@@ @@DTBO@@
```
Here in the commit description you don't have "." at the end, but I can see "." on the next commit description. please fix commit format

oleksiimoisieiev commented:
+1 `Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

GrygiriiS commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-xen.dtso:
```diff
@@ -60,3 +60,5 @@
 &pcie_rescal    { xen,passthrough; };
 &mip0           { xen,passthrough; };
+&rp1_usb0       { xen,passthrough; };
+&rp1_usb1       { xen,passthrough; };
```
There is an issue - if MMC is used by Dom0 it should not have xen,passthrough; property, otherwise Dom0 will crash with XEN IO exception.

dsemenets replied:
fixed

GrygiriiS commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-mmc.dtso:
```diff
@@ -0,0 +1,83 @@
+
+#define PHANDLE_GIC (65000)
+
+&passthrough {
```
add here to avoid build warnings/errors:
```
#address-cells = <0x2>;
#size-cells = <0x2>;
```

dsemenets replied:
Why? Those parameters present in `bcm2712-raspberrypi5-domd.dts`
I don't see any reason to add this here

GrygiriiS commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-usb.dtso:
```diff
@@ -0,0 +1,41 @@
+/dts-v1/;
+/plugin/;
+
+&rp1 {
```
add here to avoid build warnings/errors:
```

#address-cells = <0x2>;
#size-cells = <0x2>;
```

also add `interrupt-parent = <&rp1>;`

dsemenets replied:
`interrupt-parent = <&rp1>;` presented in `bcm2712-raspberrypi5-domd.dts`

lorc commented on recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:
```diff
@@ -35,6 +39,12 @@ DOM0_BOOTARGS ?= "console=hvc0 earlycon=xen earlyprintk=xen clk_ignore_unused ro
 
 do_compile() {
+    for dtbo in ${DOMD_OVERLAYS}; do
+        cat ${WORKDIR}/boot.cmd.dtapply | sed "s/@@DTBO@@/$dtbo/g" >> ${WORKDIR}/${TEMPLATE_FILE}
```
Oh wow. Why not just
```
    for dtbo in ${DOMD_OVERLAYS}; do
    	echo "fatload ${BOOT_MEDIA }0 ${XEN_DTBO_ADDR} ${dtbo}" >> ${WORKDIR}/${TEMPLATE_FILE}
	echo "fdt apply ${XEN_DTBO_ADDR}" >> ${WORKDIR}/${TEMPLATE_FILE}
    done
```
?

dsemenets replied:
Yes. Sure.

oleksiimoisieiev commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-xen.dtso:
```diff
@@ -60,3 +59,5 @@
 &pcie_rescal    { xen,passthrough; };
 &mip0           { xen,passthrough; };
+&rp1_usb0       { xen,passthrough; };
+&rp1_usb1       { xen,passthrough; };
```
So usb are passthrough by default for both cases? But when mmc is configured - these devices are not present in domd DT

dsemenets replied:
fixed

dsemenets committed:
linux: Conditional set usb controller as passthrough

If we need usb in dom0 it should not be passthrough device.

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>

dsemenets committed:
linux: Remove unused properties

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>

oleksiimoisieiev commented on recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:
```diff
@@ -9,8 +9,13 @@ INHIBIT_DEFAULT_DEPS = "1"
+    file://boot.cmd.xen.3.in \
+"
+MMC_PASSTHROUGH_DTBO = "mmc-passthrough.dtbo"
+USB_PASSTHROUGH_DTBO = "mmc-passthrough.dtbo"
```
why usb and mcc are the same?

dsemenets replied:
Misprint. Sorry

oleksiimoisieiev commented on recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb:
```diff
@@ -35,6 +40,23 @@ DOM0_BOOTARGS ?= "console=hvc0 earlycon=xen earlyprintk=xen clk_ignore_unused ro
 DOMD_BOOTARGS ?= "console=ttyAMA0 earlycon=xen earlyprintk=xen clk_ignore_unused root=\/dev\/mmcblk0p2 rootfstype=ext4 rootwait"
```
you don't change bootargs for domd

dsemenets replied:
Should I?

GrygiriiS commented:
```
Acked-by: Grygorii Strashko <grygorii_strashko@epam.com>
Tested-by:  Grygorii Strashko <grygorii_strashko@epam.com>
``` 
I'm merging this.
